Make creator cohotrs in Notions

AlexPunches opened the pull request:
в `.env` добавилась константа `NOTION_PAGE_ID=`
Это родительская страница ноушена, на которой создаются `Databases` когорт (календари)

Собстно, чтобы создать когорту, отправляем ее название на ендпоинт /cohorts/
```
{
    "name": "38 когорта. Python-разработчик"
}
```
это создаст когорту в базе, 
тутже создаст календарь в ноушене и получит его `UUID`,
и добавит `UUID` календаря к когорте в бд.

Код вроде как работает,
Но мне кажется можно его сделать красивее :)

Как тут можно улучшить?


**ЗЫ**
в прошлой версии миграций был небольшой косяк, возможно добавление когор пару раз выдаст ошибку 500,
сейчас миграции без ошибки

**ЗЗЫ**
функция `is_databases_by_cohort_name()` в `notion.py`
нормально ли использовать такие конструкции?
`db['title'][0]['text']['content']`

AlexPunches committed:
Add CBV

Ендпоинты когорт и юзеров теперь на cbv из FastAPI Utilities
роутер теперь InferringRouter() из FastAPI Utilities

AlexPunches committed:
update requirements.txt for email-validator

AlexPunches committed:
fix migrations

AlexPunches committed:
add .env NOTION_PAGE_ID and update README.md

AlexPunches committed:
change type few field

AlexPunches committed:
Make creater databas-calendar for Cohort in Notion

Теперь ендпоин /cohorts создает Когорту не только в БД,
но и в Ноушене.
При добавлении есть проверки на дубли и в БД и Ноушене

AntonZelinsky commented on app/routers/cohorts.py:
```diff
@@ -1,35 +1,46 @@
+              self,
+              cohort: schemas.CohortCreate,
+    ):
+        db_cohort = services.get_cohort_by_name(self.db, name=cohort.name)
```
Сделай метод в сервисе, который будет возвращать труфолс, незачем получать всю строку

AntonZelinsky commented on app/routers/users.py:
```diff
@@ -1,34 +1,36 @@
+            raise HTTPException(
+                status_code=400,
+                detail="User already registered",
+            )
```
Вынеси это тоже в отдельный метод, и просто используй его здесь. 
Но этим методом будет пользоваться бот, а там ловить эксепшены плохой вариант. ну, пусть будет пока, однако когда дойдём до создания пользователя в боте, подумаем как соблюсти **принцип идемпотентности**

AntonZelinsky commented on migrations/versions/2021-10-26-11-34-08_43315bdbe41f_uniq_notion_id.py:
```diff
@@ -0,0 +1,28 @@
+"""uniq-notion-id
```
чувствую, будут конфликты в миграциях, так как Влад снёс всё и создал по новой

AlexPunches replied:
надо будет на мои заменить, 
они актуальные, 
если я в них еще не успел накосячить :)

AlexPunches committed:
Change Cohorts-adder for Cohorts-Sync from Notion

AlexPunches commented:
заменил мезанизмы добавления когорт, на механизмы синхронизации из ноушена.

пока повесил синхронизацию на эндпоинт GET /cohorts/sync/,
потом уже решим как будем запускать ее

AlexPunches committed:
Remove sync

AlexPunches commented:
удалил синхронизацию, 
сейчас когорты добавляются пост-запросом на эндпоинт `/cohorts/`

AntonZelinsky commented on app/services.py:
```diff
@@ -4,16 +4,19 @@
-def get_user_by_telegram_id(db: Session, telegram_id: int):
-    return db.query(models.User).filter(
-        models.User.telegram_id == telegram_id).first()
+def has_user_by_telegram_id(db: Session, telegram_id: int) -> bool:
```
Я отписывался в твоём первом ПРе, но отпишусь и тут, тот закрою. Сделай пожалуйста два класса, один из которых будет работать только с юзерами, а другой с когортами

AlexPunches replied:
Что-то я туплю, не могу понять, что надо сделтаь. 

Разделил сервисы на классы, 
но мне кажется не так надо было сделать.

AntonZelinsky commented on .env.example:
```diff
@@ -4,4 +4,5 @@ NOTION_DATABASE_ID=
-DOMAIN_ADDRESS=
\ No newline at end of file
+DOMAIN_ADDRESS=
+NOTION_PAGE_ID=
```
Это значение нигде не используется

AntonZelinsky commented on app/services.py:
```diff
@@ -25,14 +28,21 @@ def get_cohorts(db: Session, skip: int = 0, limit: int = 100):
-    db_cohort = models.Cohort(name=cohort.name)
+    """Создать когорту в БД, если еще нет когорты с таким uuid."""
+    # TODO. FIX: uuid передается в виде строки, иначе какая-то проблема.
+    if has_cohort_by_uuid(db, str(cohort.notion_db_id)):
```
Приводи тогда к строе непосредственно перед запросом в базу. 
А вообще, проблема в том, что тип в базе у тебя строка, а по хорошему, должен быть юидом.
https://github.com/AntonZelinsky/ScheduleNotifierForMentorsOfYandex.Praktikum/pull/13/files#diff-880083c8a05dc4662af880cde1cf569e18f4476182e8680d2cee4cf1281a3d6eR26

AlexPunches replied:
исправил в базе на uuid, 
но пришлось менять первую миграцию, 
не сомг сам написать миграцию для изменения поля на uuid :(

AntonZelinsky commented on app/services.py:
```diff
@@ -25,14 +28,21 @@ def get_cohorts(db: Session, skip: int = 0, limit: int = 100):
+    )
     db.add(db_cohort)
     db.commit()
     db.refresh(db_cohort)
```
мне кажется, рефреш делать не обязательно, он сам обновит модель при комите, хотя, может я чего-то не знаю

AlexPunches replied:
так и есть, обновляется сразу после коммита.

AntonZelinsky commented on core/models.py:
```diff
@@ -22,5 +22,8 @@ class Cohort(Base):
-    name = Column(String)
-    notion_db_id = Column(String)
+    name = Column(String, unique=True)
+    notion_db_id = Column(String, unique=True)
```
если в ноушене uuid, то тут тоже лучше поменять

AntonZelinsky commented on notion.py:
```diff
@@ -89,3 +89,23 @@ def group_by_user_raw_data(raw_data: list) -> dict:
     return users_group_data
 
+
+def query_databases_by_str(query: str):
```
Если этот метод больше не используется, то по хорошему, нужно удалить. Так как этот метод заточен только под получение страниц

AntonZelinsky commented on app/routers/cohorts.py:
```diff
@@ -1,35 +1,31 @@
+    def create_cohort(self, cohort: schemas.CohortCreate):
+        try:
+            cohort = services.create_cohort(db=self.db, cohort=cohort)
+        except ValueError as e:
```
Тут не нужно перехватывать исключение. Нужно перед созданием когорты вызвать метод `has_cohort_by_uuid` и если есть, то вернуть сообщение об ошибке, можно даже с кодом 400, если нет, то создавать. В методе проверка повторится, если там упадёт, тогда такая судьба.

AlexPunches replied:
поправил, 
тепрь рейзят только сервисные функции

AlexPunches committed:
Fix few bugs

удлалена NOTION_PAGE_ID из env
в базе теперь uuid
больше нет необязательно рефреша
удалены лишние эксепшены

AlexPunches committed:
fix uuid into create_cohort()

AlexPunches committed:
add classes UserService and CohortService

AlexPunches commented:
Обновил миграции. 
Нужно было изменить тип столбца. Но Алимбик это как-то автоматом не умеет. 
А с raw при uuid не так просто было разобраться :(
Поэтому я просто запилил новые миграции с нуля.

Это значит теперь старую базу надо удалить, и накатить миграции с нуля.
Ну т.е. удалить прям volum докера, и он сам сделает новую пустую базу.

AntonZelinsky commented on app/services.py:
```diff
@@ -1,39 +1,54 @@
+        """Создать юзера в БД, если еще нет юзера с таким телеграм_ид."""
+        if cls.has_user_by_telegram_id(db, telegram_id=user.telegram_id):
+            raise HTTPException(status_code=400, detail='User already registered')
+        db_user = models.User(name=user.name, telegram_id=user.telegram_id)
```
А попробуй создание пользователя заменить такой конструкцией, по идее, они идентичные `models.User(**user)`

AlexPunches replied:
чет не получилось :(
я елси честно пока не очень понимаю пиновскую магию звездочек и других сахаров :)

что-то надо подправить наверно, пидантиковскую штуку не распаковывает

AntonZelinsky replied:
Окей, тут важно, чтобы параметры совпадали, если что-то не так, то работать не будет.

AlexPunches committed:
make Classes as Dependencies. UserService and CohortService

AntonZelinsky commented on app/routers/users.py:
```diff
@@ -1,34 +1,26 @@
+class UserCBV:
+
+    @router.get("/", response_model=List[schemas.User])
+    def read_users(self, service: UserService = Depends()):
```
Супер, давай только ещё сделаем так, чтобы сервис инжектался в ласс, а не в метод. Так код будет намного красивее.
Ты уже делал такое в UserService или вот пример как сделать красиво https://fastapi-utils.davidmontague.xyz/user-guide/class-based-views/#the-cbv-decorator

AlexPunches replied:
с этим вроде разобрался

AntonZelinsky commented on app/services.py:
```diff
@@ -1,39 +1,65 @@
+                 ):
+        self.db = db
+        self.skip = skip
+        self.limit = limit
```
Как я понимаю, эти параметры используются только для функции get_users, к тому же, они туда инжектаются напрямую. 
Я предлагаю их отсюда удалить, впрочем как и получении сессии и сделать как я привёл в примере выше. Так код будет лаконичнее и красивее.

AlexPunches replied:
вроде сделал как смог понять )
небольшой вопрос остался, перенесу его в слак.

AlexPunches committed:
fix int and limit

AlexPunches committed:
move Session to cbv

AlexPunches committed:
move Session to Services

AntonZelinsky commented on app/services.py:
```diff
@@ -1,39 +1,56 @@
-        models.User.telegram_id == telegram_id).first()
 
+class Services:
+    def __init__(self, db: Session = Depends(get_db)):
```
Смысла от этой абстракции не очень много, но раз сделал, то пусть будет.
